pythonwhat/tasks.py

```diff
@@ -362,41 +362,73 @@ def taskRunEval(
+    Utility function later wrapped to extract either result (end state of a variable), output (stdout) or error
+
+    Args:
+        tree (ast): current focused ast, used to get code to execute
+        process: manages shell (see locals.py)
+        shell: link to to get process namespace from execution up until now
+        env: variables to be replaced in focused code by name
+        extra_env: variables to be replaced in focused code by name from extra_env in has_expr
+        context: sum of set_context in sct chain
+        context_vals: extra context argument in has_expr
+        pre_code: argument in has_expr to execute code before evaluating, for example to set a seed
+        expr_code: code to execute instead of focused code
+        name: extract value after executing focused expr_code (~post_code)
+        copy: copy entire env because our expr_code could have side effects
+        tempname: key for the result when it is added to context, only for v1 sct's
+        call: only used in v1 sct's
+
+    Returns:
+        str: output of the executed code
+    """
     try:
         # Prepare code and mode -----------------------------------------------
-        if (expr_code and name) or (not expr_code and isinstance(tree, ast.Module)):
+        # Verify if expr_code is expression code (returning a value) or just runnable code.
+        if (  # expr_code returns nothing and then we will extract a value
+            expr_code and name
+        ) or (  # No expr_code and the tree is of a node type that does not evaluate to have output
+            not expr_code and isinstance(tree, ast.Module)
+        ):
+            # We are not focused on an expression (no output)
             mode = "exec"
         else:
             mode = "eval"
+            # Wrap the focused node in the tree so it can be run with eval()
             if not isinstance(tree, (ast.Module, ast.Expression, ast.Expr)):
                 tree = ast.Expression(tree)
 
         # Expression code takes precedence over tree code
         if expr_code:
             code = expr_code
             tree = ast.parse(code, mode=mode)
-        else:
+        else:  # Compile the tree to Python code
             code = compile(tree, "<script>", mode)
 
         # Set up environment --------------------------------------------------
-        # avoid deep copy if specified, or if just looking up variable by name
-        # unpack 'container nodes' first
+        # Unpack 'container nodes' before checking if a deepcopy is needed
         if isinstance(tree, ast.Module):
             tree = tree.body
         if isinstance(tree, ast.Expression):
             tree = tree.body
         if isinstance(tree, ast.Expr):
             tree = tree.value
+
+        # Avoid a deep copy if specified or if the ast node type indicates we are looking up a variable by name
+        # ast.Name, ast.Subscript and ast.Load do not have side effects in the environment so a deepcopy is not needed
```

Comment:
(most of the time, hence the possibility to override: custom getters could have side effects, e.g a usage counter)

Reply:
Fixed